### PR TITLE
Issue635 partial classes

### LINKED ISSUE #635: ArgumentException in DisposeFieldsAndPropertiesInTearDownAnalyzer
```
Analyzer 'NUnit.Analyzers.DisposeFieldsInTearDown.DisposeFieldsAndPropertiesInTearDownAnalyzer' threw an exception of type 'System.ArgumentException' with message 'Syntax node is not within syntax tree'
```

```
1>    SyntaxNode: public partial class TestFixture ... [ClassDeclarationSyntax]@[17..158) (2,0)-(8,1)
1>    System.ArgumentException: Syntax node is not within syntax tree
1>    at Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.CheckSyntaxNode(CSharpSyntaxNode syntax)
1>    at Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.GetSymbolInfo(ExpressionSyntax expression, CancellationToken cancellationToken)
1>    at Microsoft.CodeAnalysis.CSharp.CSharpExtensions.GetSymbolInfo(SemanticModel semanticModel, ExpressionSyntax expression, CancellationToken cancellationToken)
1>    at NUnit.Analyzers.DisposeFieldsInTearDown.DisposeFieldsAndPropertiesInTearDownAnalyzer.Parameters.IsLocalMethodCall(InvocationExpressionSyntax invocationExpression, IMethodSymbol& calledMethod)
1>    at NUnit.Analyzers.DisposeFieldsInTearDown.DisposeFieldsAndPropertiesInTearDownAnalyzer.AssignedIn(Parameters parameters, HashSet`1 assignments, ExpressionSyntax expression)
1>    at NUnit.Analyzers.DisposeFieldsInTearDown.DisposeFieldsAndPropertiesInTearDownAnalyzer.AssignedIn(Parameters parameters, HashSet`1 assignments, SyntaxList`1 statements)
1>    at NUnit.Analyzers.DisposeFieldsInTearDown.DisposeFieldsAndPropertiesInTearDownAnalyzer.AssignedIn(Parameters parameters, HashSet`1 assignments, BaseMethodDeclarationSyntax method)
1>    at NUnit.Analyzers.DisposeFieldsInTearDown.DisposeFieldsAndPropertiesInTearDownAnalyzer.AssignedIn(Parameters parameters, HashSet`1 assignments, IMethodSymbol symbol)
1>    at NUnit.Analyzers.DisposeFieldsInTearDown.DisposeFieldsAndPropertiesInTearDownAnalyzer.AssignedIn(Parameters parameters, HashSet`1 assignments, IEnumerable`1 methods)
1>    at NUnit.Analyzers.DisposeFieldsInTearDown.DisposeFieldsAndPropertiesInTearDownAnalyzer.AnalyzeAssignedButNotDisposed(SyntaxNodeAnalysisContext context, Dictionary`2 symbols, Parameters parameters, String where, IEnumerable`1 setUpMethods, IEnumerable`1 tearDownMethods, HashSet`1 assignedWithInitializers)
1>    at NUnit.Analyzers.DisposeFieldsInTearDown.DisposeFieldsAndPropertiesInTearDownAnalyzer.AnalyzeDisposableFields(SyntaxNodeAnalysisContext context)
1>    at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.<>c__49`1.<ExecuteSyntaxNodeAction>b__49_0(ValueTuple`2 data)
1>    at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteAndCatchIfThrows_NoLock[TArg](DiagnosticAnalyzer analyzer, Action`1 analyze, TArg argument, Nullable`1 info, CancellationToken cancellationToken)

```

NUnit: 3.14.0
NUnit.Analyzers: 3.9.0

```cs
using System;

public partial class TestFixture
{
    private static object Property
    {
        get => throw new NotImplementedException();
    }
}

```

```cs
using NUnit.Framework;
using System;

public partial class TestFixture
{
	[Test]
	public void SomeTest()
	{
		SomeAsserts();
	}

	private static void SomeAsserts() => throw new InvalidOperationException();
}

```

### LINKED ISSUE #659: `NUnit1032` throws an exception in a specific configuration
The following code fails:

``` csharp
using NUnit.Framework;

using System.Diagnostics.CodeAnalysis;

namespace NUnitRepro;

[TestFixture]
public abstract class Test : IDataProvider
{
    private readonly DataReader DataReader;

    DataReader IDataProvider.DataReader
    {
        [RequiresUnreferencedCode("Reflection-based JSON")]
        get => DataReader;
    }

    [RequiresUnreferencedCode("Reflection-based JSON")]
    protected Test() => DataReader = new DataReader(this);

    [OneTimeSetUp] public void SetListener() => throw new NotImplementedException();
}

public interface IDataProvider
{
    DataReader DataReader
    {
        [RequiresUnreferencedCode("Reflection-based JSON")]
        get;
    }
}

[RequiresUnreferencedCode("Reflection-based JSON")]
public sealed class DataReader
{
    public DataReader(IDataProvider provider) => throw new NotImplementedException();
}
```

This gives the following error:

``` text
error AD0001: Analyzer 'NUnit.Analyzers.DisposeFieldsInTearDown.DisposeFieldsAndPropertiesInTearDownAnalyzer' threw an exception of type 'System.ArgumentException' with message 'An item with the same key has already been added. Key: DataReader'
```

If you remove `abstract`, or the attribute on the getter of the explicit implementation of `IDataProvider.DataReader` on `Test`, or the `[OneTimeSetUp]` attribute, this starts to work.

### PR DESCRIPTION
**Pull request by manfred-brands**
Fixes #635 
Fixes #659

### CODE REVIEW

**Review by mikkelbu — APPROVED** [review]

File: src/nunit.analyzers/DisposeFieldsAndPropertiesInTearDown/DisposeFieldsAndPropertiesInTearDownAnalyzer.cs [hunk 1]
```diff
@@ -88,25 +88,49 @@ private static void AnalyzeDisposableFields(SyntaxNodeAnalysisContext context)
...
+                {
+                    if (NeedsDisposal(model, property.Initializer.Value))
+                    {
+                        symbolsWithDisposableInitializers.Add(property.Identifier.Text);
+                    }
+                }
+
+                if (CanBeAssignedTo(property))
+                {
+                    symbols.Add(property.Identifier.Text, property);
```
[the remark this answers is not in the document]
Comment (manfred-brands) [reply src/nunit.analyzers/DisposeFieldsAndPropertiesInTearDown/DisposeFieldsAndPropertiesInTearDownAnalyzer.cs:126]: Thanks @mikkelbu for noticing this.

A field with no _initializer_, can always be assigned to, even `readonly` fields can be assigned in a constructor.
A field with an _initalizer_, could be a `const`/`readonly` field, only the latter would need checking for disposal.

A property without an _initializer_ could still a `get` only one.
But I see that I made two mistakes:
1. If the property needs disposal it wasn't added to the `symbols` list.
2. Just like `readonly` fields, `get` only properties can be assigned to in constructor.

I need to add a few more unit tests and update code accordingly.

[the remark this answers is not in the document]
Comment (manfred-brands) [reply src/nunit.analyzers/DisposeFieldsAndPropertiesInTearDown/DisposeFieldsAndPropertiesInTearDownAnalyzer.cs:130]: @mikkelbu The sections for _field_ and _Property_ are now virtually the same.
I added a check that if there is some constructor, then readonly fields and get-only properties can be assigned to and hence need checking.

**Review by mikkelbu — APPROVED** [review]
LGTM. Thanks for all the work @manfred-brands

**Comment by alteredego** [comment]
I'm facing the same exception being thrown. Any chance this fix could be rolled out into a 3.11 release?

**Comment by mikkelbu** [comment]
> I'm facing the same exception being thrown. Any chance this fix could be rolled out into a 3.11 release?

@alteredego I'll will see if I can get a release out this weekend - either a 3.11 or a 4.00

**Comment by mikkelbu** [comment]
We will not backport the fix to 3.11, but the NUnit 4.X analyzers should work with both NUnit 3 and NUnt 4.

**Comment by alteredego** [comment]
I'll give 4.0.1 a go. Many thanks!

**Pull request merged**